### PR TITLE
migrate documentation to Antora (jetty-10.0.x)

### PR DESCRIPTION
**Pull request by mojavelinux**

### CODE REVIEW

File: documentation/jetty/README.adoc [hunk 1]
```diff
@@ -0,0 +1,17 @@
+= Jetty Documentation
```
Comment (sbordet) [inline documentation/jetty/README.adoc:14]: We have a license header rule fo r`*.adoc` files.

Either add the license header file here, or you can exclude it in the top-level POM, around line 1928.

I prefer the latter.

File: tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.tests.hometester;
+
+import java.nio.file.Files;
```
Comment (sbordet) [inline tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java]: Unused import.

File: tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+import java.nio.file.Path;
+import java.nio.file.Paths;
```
Comment (sbordet) [inline tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java]: Unused import.

File: tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    {
+        Map<String, String> config = new HashMap<String, String>();
+        String name = null;
+        for (String arg : args)
+        {
+          if (arg.startsWith("--") && arg.contains("="))
+          {
+              String[] nameAndValue = arg.split("=", 2);
+              config.put(nameAndValue[0].substring(2), nameAndValue[1]);
+          }
```
Comment (sbordet) [inline tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java]: These lines have the wrong indentiation, which is 4 spaces.

File: tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+        }
+        if (!config.containsKey("jetty-home")) throw new RuntimeException("--jetty-home argument is required");
```
Comment (sbordet) [inline tests/jetty-home-tester/src/main/java/org/eclipse/jetty/tests/hometester/RunJetty.java]: The `throw` statement must be on the next line and indented.

**Review by sbordet — CHANGES_REQUESTED** [review]

File: documentation/jetty/README.adoc (hunk 1 above)
Comment (mojavelinux) [reply documentation/jetty/README.adoc:14]: I'll just add it here since we are adding it for all other .adoc files. I just forgot about this one.

**Comment by sbordet** [comment]
@mojavelinux I could build the documentation, but the `plantuml` diagrams are not rendered.

I do have `dot` installed on my system.

For example:
https://webtide.github.io/jetty.website/docs/jetty/10/operations-guide/protocols/index.html#http2

Before and after the paragraph starting with `Nowadays` there are 2 diagrams, but they are not rendered as they make an external request to `kroki.io`?

For the locally built documentation, the diagram source code is rendered instead.

Can we build the diagrams locally, store the images locally, and render them in the documentation as local image files?

**Comment by mojavelinux** [comment]
Generating the diagrams is handled by Kroki via a remote requests. Relying on Kroki ensures you don't have to have any diagram tools installed on your machine. But the local branch preview does not load asciidoctor-kroki in order to make the preview as fast as possible. It could be added though. To do so, the playbook template at https://github.com/jetty/jetty.website/blob/main/lib/playbook-templates/per-branch-antora-playbook.yml must be modified.

First, asciidoctor-kroki needs to be added to the package line. Then, asciidoctor-kroki needs to be added to the list of asciidoc.extensions. If that is what you prefer, we can proceed with that change.

**Comment by sbordet** [comment]
I prefer the site to not perform any external request.

Last time I tried https://webtide.github.io/jetty.website/, kroki.io was down or super-slow, so the site kinda hung, and the diagrams were not there, so bad experience.

I can live with previews or local to remotely call kroki.io, but the "production" site should have local images and not do any external request.

We already require a bunch of tools to build the website anyway, so one more is ok.

Please proceed with the change. Thanks!

**Comment by mojavelinux** [comment]
> I can live with previews or local to remotely call kroki.io, but the "production" site should have local images and not do any external request.

Currently, there's no other option when using Antora unless you set up your own kroki server. Even then, Antora has to make calls to the kroki URL (whether it is your own or the shared one) to generate diagrams.

**Comment by mojavelinux** [comment]
> Please proceed with the change. Thanks!

See https://github.com/jetty/jetty.website/pull/27

**Comment by sbordet** [comment]
> Currently, there's no other option when using Antora unless you set up your own kroki server. Even then, Antora has to make calls to the kroki URL (whether it is your own or the shared one) to generate diagrams.

I'm not sure where this limitation come from?
Asciidoctor can generate the diagrams locally, so I would imagine Antora would just delegate to it?

**Comment by mojavelinux** [comment]
> I'm not sure where this limitation come from?
> Asciidoctor can generate the diagrams locally, so I would imagine Antora would just delegate to it?

Asciidoctor cannot generate diagrams on its own. It relies on extensions to do that. When using Asciidoctor (Ruby) or AsciidoctorJ, that responsibility is handled by Asciidoctor Diagram. Asciidoctor Diagram generates diagrams using local tools. However, Asciidoctor Diagram is not available for Asciidoctor.js, which is what Antora uses. That's why in Antora, we use Asciidoctor Kroki. The reason Asciidoctor Kroki works across all environments is because it delegates to a service (the Kroki service) to generate the diagrams. This is currently the only way to generate diagrams when using Antora.

**Comment by mojavelinux** [comment]
Given how Kroki is configured by default, the build won't make any requests to Kroki.io. It will only happen when the page that contains the diagram is viewed. It's a remote image.

**Review by gregw — CHANGES_REQUESTED** [review]
@mojavelinux @jmcc0nn3ll  Sorry if I'm late to the review party... but I've just gone through the new website as running https://jetty.github.io/jetty.website/ and here is some feedback:

 + The home page has contents in the top right, whilst the Documentation pages have the contents on the left boundary.    We should be consistent and I much prefer the left boundary approach.....
    * Oh I see, there are also Contents pages on the right of the Documentation, but only on pages that have sections.      So the left is a menu and the right is page contents?   Maybe we need a left Menu on the home page for Documentation, Github, Eclipse, Webtide etc.

+ There are no links out from the homepage to: the blogs; eclipse; github; webtide; etc.  There is one buried link to the documentation, but it is not very prominent.    I don't think we need a paragraph saying how good our blogs are, we just need an entry in the Contents saying "Jetty Blogs" that is a link directly to the blog page.  Likewise we need prominent links in the Contents to Documentation, Gitub project, Webtide.   The home page is our landing page and it has to be easy to navigate from.... in fact the primary purpose of that page should be to navigate to other pages... few will read the text.

+ On the front page of the Jetty Documentation, in the resources column can we add links to issues

+ The documentation pages do not use wide windows. All the text is in a narrow column with lots of whitespace either side.

 + On a mobile device, it is kind of strange that the left menu is collapsed, by the the page contents is always listed.    Can we make the Contents display collapsed on a small device, as often you only need the contents links to skip the contents that fills the first page.

+ The titles for the top level pages of Jetty-10, Jetty-11 and Jetty-12 should have the version number in the title. e.g on https://jetty.github.io/jetty.website/docs/jetty/10/index.html it is only apparently we are looking at Jetty-10 by the small selector top right and by reading the body of the first line.   Put it in the title, so it will appear on the left menu, top of the page and tab/page title.

**Comment by mojavelinux** [comment]
This not the correct place to be providing feedback. This is an automated PR generated by the migration. Discussions should be directed to the issues in the jetty.website repository.

**Review by sbordet — APPROVED** [review]

**Pull request merged**